File: 백준/Gold/2470. 두 용액/두 용액.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		int N=Integer.parseInt(br.readLine());
		StringTokenizer st=new StringTokenizer(br.readLine());

		int[] map=new int[N];



		for(int i=0;i<N;i++){
			map[i]=Integer.parseInt(st.nextToken());
		}

		Arrays.sort(map);

		int start=0;
		int end=N-1;
		int gap=Integer.MAX_VALUE;
		int temp1=0,temp2=0;

		while(start<end){
			int sum=map[start]+map[end];
			if(Math.abs(sum)<gap){
				gap=Math.abs(sum);
				temp1=map[start];
				temp2=map[end];
			}
			if(sum>0){
				end=end-1;
			}else{
				start=start+1;
			}
		}

		System.out.println(temp1+" "+temp2);


	}
}

File: 백준/Gold/2470. 두 용액/두 용액.py
import sys
input=sys.stdin.readline

N=int(input())

liquids=list(map(int,input().split()))

liquids.sort()

def binarySearch():
    minGap=2000000001
    a=0
    b=0
    start=0
    end=N-1

    while(start<end):
        gap=liquids[start]+liquids[end]
        if minGap>abs(gap):
            minGap=abs(gap)
            a=liquids[start]
            b=liquids[end]

        if gap<0:
            start+=1
        elif gap>0:
            end-=1
        else: break;

    print(f"{a} {b}")

if liquids[N-1]<0:
    print(f'{liquids[N-2]} {liquids[N-1]}')
elif liquids[0]>=0:
    print(f'{liquids[0]} {liquids[1]}')
else:
    binarySearch()

